index/segment/mem/simple_terms_dict.go

```diff
@@ -0,0 +1,212 @@
+	return dict
+}
+
+func (t *simpleTermsDictionary) Insert(field doc.Field, i segment.DocID) error {
+	fieldName := string(field.Name)
+	fieldValues := t.getOrAddFieldName(fieldName)
+	fieldValue := string(field.Value)
+	return fieldValues.addDocIDForValue(fieldValue, i)
+}
+
+func (t *simpleTermsDictionary) Fetch(
+	fieldName []byte,
+	fieldValueFilter []byte,
+	opts termFetchOptions,
+) (segment.PostingsList, error) {
+	// check if we know about the field name
+	t.fields.RLock()
+	fieldValues, ok := t.fields.nameToValuesMap[string(fieldName)]
+	t.fields.RUnlock()
+	if !ok {
+		// not an error to not have any matching values
+		return nil, nil
+	}
+
+	// get postingList(s) for the given value.
+	lists, err := fieldValues.fetchLists(fieldValueFilter, opts.isRegexp)
+	if err != nil {
+		return nil, err
+	}
+
+	// union all the postingsList(s)
+	unionedList := t.opts.PostingsListPool().Get()
+	for _, ids := range lists {
+		unionedList.Union(ids)
+	}
+
+	return unionedList, nil
+}
+
+func (t *simpleTermsDictionary) getOrAddFieldName(fieldName string) *fieldValuesMap {
+	// cheap read lock to see if it already exists
+	t.fields.RLock()
+	fieldValues, ok := t.fields.nameToValuesMap[fieldName]
+	t.fields.RUnlock()
+	if ok {
+		return fieldValues
+	}
+
+	// acquire write lock and create
+	t.fields.Lock()
+	fieldValues, ok = t.fields.nameToValuesMap[fieldName]
+
+	// check if it's been created since we last acquired the lock
+	if ok {
+		t.fields.Unlock()
+		return fieldValues
+	}
+
+	fieldValues = newFieldValuesMap(t.opts)
+	t.fields.nameToValuesMap[fieldName] = fieldValues
+	t.fields.Unlock()
+	return fieldValues
+}
+
+type fieldValuesMap struct {
+	sync.RWMutex
+
+	opts Options
+
+	// fieldValue -> postingsList
+	values map[string]segment.PostingsList
+	// TODO: as noted in https://github.com/m3db/m3ninx/issues/11, evalute impact of using
+	// a custom hash map where we can avoid using string keys, both to save allocs and
+	// help perf.
+}
+
+func newFieldValuesMap(opts Options) *fieldValuesMap {
+	return &fieldValuesMap{
+		opts:   opts,
+		values: make(map[string]segment.PostingsList),
+	}
+}
+
+func (f *fieldValuesMap) addDocIDForValue(value string, i segment.DocID) error {
+	// try read lock to see if we already have a postingsList for the given value.
+	f.RLock()
+	pid, ok := f.values[value]
+	f.RUnlock()
+
+	// we have a postingsList, mark the docID and move on.
+	if ok {
+		pid.Insert(i)
+		return nil
+	}
+
+	// postingsList doesn't exist, time to acquire write lock
+	f.Lock()
+	pid, ok = f.values[value]
+
+	// check if it's been created since we released lock
+	if ok {
+		f.Unlock()
+		pid.Insert(i)
+		return nil
+	}
+
+	// create new posting id for the term, and insert into fieldValues
+	pid = f.opts.PostingsListPool().Get()
+	f.values[value] = pid
+	f.Unlock()
+	pid.Insert(i)
+	return nil
+}
+
+func (f *fieldValuesMap) fetchLists(valueFilter []byte, regexp bool) ([]segment.PostingsList, error) {
```

Comment:
At some point we might like to turn this into an iterator, to avoid the array alloc for returning the result, etc.  Is tough when using a map backed structure for the field values map however.   

Just thought I'd mention it, this is good for now though.

Reply:
Yep sounds good.

Reply:
done.